Mathlib/Combinatorics/SimpleGraph/Triangle/Tripartite.lean

```diff
@@ -0,0 +1,244 @@
+      (graph t).Adj (in₀ a) (in₂ c) ∧ (graph t).Adj (in₁ b) (in₂ c) := by
+  rintro (_ | _ | _) (_ | _ | _) (_ | _ | _) <;>
+    refine ⟨_, _, _, by ext; simp only [Finset.mem_insert, Finset.mem_singleton]; try tauto,
+      ?_, ?_, ?_⟩ <;> constructor <;> assumption
+
+/-- The map that turns a triangle index into an explicit triangle. -/
+@[simps] def toTriangle : α × β × γ ↪ Finset (α ⊕ β ⊕ γ) where
+  toFun x := {in₀ x.1, in₁ x.2.1, in₂ x.2.2}
+  inj' := fun ⟨a, b, c⟩ ⟨a', b', c'⟩ ↦ by simpa only [Finset.Subset.antisymm_iff, Finset.subset_iff,
+    mem_insert, mem_singleton, forall_eq_or_imp, forall_eq, Prod.mk.inj_iff, or_false, false_or,
+    in₀, in₁, in₂, Sum.inl.inj_iff, Sum.inr.inj_iff] using And.left
+
+lemma toTriangle_is3Clique (hx : x ∈ t) : (graph t).IsNClique 3 (toTriangle x) := by
+  simp only [toTriangle_apply, is3Clique_triple_iff, in₀₁_iff, in₀₂_iff, in₁₂_iff]
+  exact ⟨⟨_, hx⟩, ⟨_, hx⟩, _, hx⟩
+
+lemma exists_mem_toTriangle {x y : α ⊕ β ⊕ γ} (hxy : (graph t).Adj x y) :
+    ∃ z ∈ t, x ∈ toTriangle z ∧ y ∈ toTriangle z := by cases hxy <;> exact ⟨_, ‹_›, by simp⟩
+
+nonrec lemma is3Clique_iff [NoAccidental t] {s : Finset (α ⊕ β ⊕ γ)} :
+    (graph t).IsNClique 3 s ↔ ∃ x, x ∈ t ∧ toTriangle x = s := by
+  refine ⟨fun h ↦ ?_, ?_⟩
+  · rw [is3Clique_iff] at h
+    obtain ⟨x, y, z, hxy, hxz, hyz, rfl⟩ := h
+    obtain ⟨a, b, c, habc, hab, hac, hbc⟩ := graph_triple hxy hxz hyz
+    refine ⟨(a, b, c), ?_, habc⟩
+    obtain ⟨c', hc'⟩ := in₀₁_iff.1 hab
+    obtain ⟨b', hb'⟩ := in₀₂_iff.1 hac
+    obtain ⟨a', ha'⟩ := in₁₂_iff.1 hbc
+    obtain (rfl | rfl | rfl) := NoAccidental.wow ha' hb' hc' <;> assumption
+  · rintro ⟨x, hx, rfl⟩
+    exact toTriangle_is3Clique hx
+
+lemma toTriangle_surjOn [NoAccidental t] :
+    (t : Set (α × β × γ)).SurjOn toTriangle ((graph t).cliqueSet 3) := fun _ ↦ is3Clique_iff.1
+
+variable (t)
+
+lemma map_toTriangle_disjoint [ExplicitDisjoint t] :
+    (t.map toTriangle : Set (Finset (α ⊕ β ⊕ γ))).Pairwise
+      fun x y ↦ (x ∩ y : Set (α ⊕ β ⊕ γ)).Subsingleton := by
+  clear x
+  intro
+  simp only [Finset.coe_map, Set.mem_image, Finset.mem_coe, Prod.exists, Ne,
+    forall_exists_index, and_imp]
+  rintro a b c habc rfl e x y z hxyz rfl h'
+  have := ne_of_apply_ne _ h'
+  simp only [Ne, Prod.mk.inj_iff, not_and] at this
+  simp only [toTriangle_apply, in₀, in₁, in₂, Set.mem_inter_iff, mem_insert, mem_singleton,
+    mem_coe, and_imp, Sum.forall, or_false, forall_eq, false_or, eq_self_iff_true, imp_true_iff,
+    true_and, and_true, Set.Subsingleton]
+  suffices ¬ (a = x ∧ b = y) ∧ ¬ (a = x ∧ c = z) ∧ ¬ (b = y ∧ c = z) by aesop
+  refine ⟨?_, ?_, ?_⟩
+  · rintro ⟨rfl, rfl⟩
+    exact this rfl rfl (ExplicitDisjoint.inj₂ habc hxyz)
+  · rintro ⟨rfl, rfl⟩
+    exact this rfl (ExplicitDisjoint.inj₁ habc hxyz) rfl
+  · rintro ⟨rfl, rfl⟩
+    exact this (ExplicitDisjoint.inj₀ habc hxyz) rfl rfl
+
+lemma cliqueSet_eq_image [NoAccidental t] : (graph t).cliqueSet 3 = toTriangle '' t :=
+by ext; exact is3Clique_iff
```

Comment:
The prevailing style is to put the by at the end of the line (several times)

Reply:
As you might have noticed, ths file was hand-ported from Lean 3 😁